processor/src/trace/utils.rs

```diff
@@ -190,10 +191,62 @@ impl ChipletsLengths {
     }
 }
 
+// AUXILIARY COLUMN BUILDER
+// ================================================================================================
+
+/// Defines a builder responsible for building a single column in an auxiliary segment of the
+/// execution trace.
+pub trait AuxColumnBuilder<E: FieldElement<BaseField = Felt>> {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    fn requests(&self, main_trace: &MainTrace, alphas: &[E], i: usize) -> E;
+
+    fn responses(&self, main_trace: &MainTrace, alphas: &[E], i: usize) -> E;
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    fn init_requests(&self, _main_trace: &MainTrace, _alphas: &[E]) -> E {
+        E::ONE
+    }
+
+    fn init_responses(&self, _main_trace: &MainTrace, _alphas: &[E]) -> E {
+        E::ONE
+    }
+
+    /// Builds the chiplets bus auxiliary trace column.
+    fn build_aux_column(&self, main_trace: &ColMatrix<Felt>, alphas: &[E]) -> Vec<E> {
+        let main_trace = MainTrace::new(main_trace);
+        let mut responses_prod: Vec<E> = unsafe { uninit_vector(main_trace.num_rows()) };
+        let mut requests_prod: Vec<E> = unsafe { uninit_vector(main_trace.num_rows()) };
+        responses_prod[0] = self.init_responses(&main_trace, alphas);
+        requests_prod[0] = self.init_requests(&main_trace, alphas);
+
+        let mut result_2_acc = E::ONE;
+        for i in 0..main_trace.num_rows() - 1 {
+            responses_prod[i + 1] = responses_prod[i] * self.responses(&main_trace, alphas, i);
+            requests_prod[i + 1] = self.requests(&main_trace, alphas, i);
+            result_2_acc *= requests_prod[i + 1];
+        }
+
+        let mut acc_inv = result_2_acc.inv();
+
+        for i in (0..main_trace.num_rows()).rev() {
+            responses_prod[i] *= acc_inv;
+            acc_inv *= requests_prod[i];
+        }
+        responses_prod
+    }
```

Comment:
It took me a bit to understand what this was doing, and I think we can improve it (readability- and memory-wise). The algorithm would be roughly:

```rs
// `requests_prod` is converted to a BTreeMap that remembers only the indices that had an actual request
let mut requests = BTreeMap::new();
requests.insert(0, self.init_requests());
let mut requests_running_prod = self.init_requests();

let mut responses_prod: Vec<E> = uninit_vector();
responses_prod[0] = self.init_responses();

for i in 0..main_trace.num_rows() - 1 {
  // Notice: `self.requests()` and `responses()` would now return an `Option` (`None` means "no request/response")
  responses_prod[i + 1] = responses_prod[i] * self.responses(&main_trace, alphas, i).unwrap_or(ONE);
  
  if let Some(req) = self.requests(main_trace, alphas, i) {
    requests.insert(i + 1, req);
    requests_running_prod *= req;
  }
}

let denominator_inv = requests_running_prod.inv();
let mut result_aux_column = responses;

for i in (0..main_trace.num_rows()).rev() {
  result_aux_column[i] *= denominator_inv;
  if let Some(req) = requests.get(i) {
    denominator_inv *= req;
  }
}

result_aux_column
```

The benefits I see:
- the `requests/responses()` methods return an `Option`, so it's clear from the API when a request/response was generated.
  - This means that we see in this algorithm what to do when there's no request/response (e.g. multiply by `ONE`)
- The `BTreeMap` for `requests` requires less memory than the `requests_prod` array (which is presumably sparse)
- Improved variable naming to make it easier to follow

I would also add a comment that this algorithm is an optimized version of multiplying by response and dividing by request at each row (which we don't do because `.inv()` is expensive).

Reply:
It would be interesting to see if using a `BTreeMap` is actually more efficient here. Getting a value from a vector should result in one memory read, but getting a value from a `BTreeMap` may require `log(n)` reads - right? If the map is small enough, it would probably still be faster as the entire map would be in the cache - but if there are many requests, the vector approach may be faster.

Reply:
Yes, this would optimize for memory usage but it is hard for me to compare performance. Should we implement both and compare? Even then we might need to benchmark typical trace instances to get an average estimate of performance.

Reply:
> Should we implement both and compare? Even then we might need to benchmark typical trace instances to get an average estimate of performance.

We should probably open a new issue for this and do this in another PR. It'd be good to have good benchmarks too more generally (haven't looked at what our current benchmarks actually measure yet).

For this PR though, I would then just fix variable names (will open separate comments for that)

Reply:
Created #1202